File: src/devtwin/system/process.py
"""Process inspection.

Best-effort, stdlib + platform-native tools only (no psutil dependency).
Never exposes full command lines by default -- only pid/name -- to avoid
leaking secrets that are sometimes passed as CLI arguments (e.g.
``--password=...``).
"""

from __future__ import annotations

import platform

from devtwin.core.models import ProcessInfo
from devtwin.core.runner import run_command


def list_processes_matching(name_substring: str) -> list[ProcessInfo]:
    """Find running processes whose name contains ``name_substring``."""
    system = platform.system()
    processes: list[ProcessInfo] = []
    needle = name_substring.lower()

    if system == "Windows":
        result = run_command(
            ["tasklist", "/FO", "CSV", "/NH"],
            timeout=5,
        )
        if not result.available or result.returncode != 0:
            return processes
        for line in result.stdout.splitlines():
            fields = [f.strip('"') for f in line.split('","')]
            if len(fields) < 2:
                continue
            proc_name = fields[0].strip('"')
            if needle not in proc_name.lower():
                continue
            try:
                pid = int(fields[1])
            except ValueError:
                continue
            processes.append(ProcessInfo(pid=pid, name=proc_name))
        return processes

    result = run_command(["ps", "-eo", "pid,comm"], timeout=5)
    if not result.available or result.returncode != 0:
        return processes
    for line in result.stdout.splitlines()[1:]:
        line = line.strip()
        if not line:
            continue
        parts = line.split(None, 1)
        if len(parts) != 2:
            continue
        pid_str, comm = parts
        if needle not in comm.lower():
            continue
        try:
            pid = int(pid_str)
        except ValueError:
            continue
        processes.append(ProcessInfo(pid=pid, name=comm))
    return processes

File: mcphub/devtwin/src/system/process.ts
/**
 * Process inspection.
 *
 * Best-effort, platform-native tools only (no extra dependency). Never
 * exposes full command lines by default -- only pid/name -- to avoid leaking
 * secrets that are sometimes passed as CLI arguments (e.g.
 * `--password=...`).
 */

import type { ProcessInfo } from '../core/models.js';
import { runCommand } from '../core/runner.js';
import { platformSystem } from './os.js';

/** Equivalent of Python's `str.strip('"')`: drop leading/trailing quote chars. */
function stripQuotes(value: string): string {
  return value.replace(/^"+/, '').replace(/"+$/, '');
}

/** Find running processes whose name contains `nameSubstring`. */
export async function listProcessesMatching(nameSubstring: string): Promise<ProcessInfo[]> {
  const system = platformSystem();
  const processes: ProcessInfo[] = [];
  const needle = nameSubstring.toLowerCase();

  if (system === 'Windows') {
    const result = await runCommand(['tasklist', '/FO', 'CSV', '/NH'], { timeout: 5 });
    if (!result.available || result.returncode !== 0) {
      return processes;
    }
    for (const line of result.stdout.split(/\r?\n/)) {
      const fields = line.split('","').map((f) => stripQuotes(f));
      if (fields.length < 2) {
        continue;
      }
      const procName = stripQuotes(fields[0]!);
      if (!procName.toLowerCase().includes(needle)) {
        continue;
      }
      const pid = Number.parseInt(fields[1]!, 10);
      if (Number.isNaN(pid)) {
        continue;
      }
      processes.push({ pid, name: procName, port: null });
    }
    return processes;
  }

  const result = await runCommand(['ps', '-eo', 'pid,comm'], { timeout: 5 });
  if (!result.available || result.returncode !== 0) {
    return processes;
  }
  // Skip the header row.
  for (const rawLine of result.stdout.split(/\r?\n/).slice(1)) {
    const line = rawLine.trim();
    if (!line) {
      continue;
    }
    // Split once on whitespace: "<pid> <command name>".
    const parts = /^(\S+)\s+(.*)$/.exec(line);
    if (!parts) {
      continue;
    }
    const pidStr = parts[1]!;
    const comm = parts[2]!;
    if (!comm.toLowerCase().includes(needle)) {
      continue;
    }
    const pid = Number.parseInt(pidStr, 10);
    if (Number.isNaN(pid)) {
      continue;
    }
    processes.push({ pid, name: comm, port: null });
  }
  return processes;
}
